File: algorithms/0-1 Knapsack Problem/Knapsack.cpp
/**
 * 0-1 Knapsack using DP
 * */
#include<iostream>
#include<algorithm>
using namespace std;

int knap(int N,int W,int wt[],int val[]);

int main(){
    int w[3]={10,20,30};
    int val[3]={60,100,120};
    cout<<knap(3,50,w,val);
}

int knap(int N,int W,int wt[],int val[]){
    int k[N+1][W+1];
    int i,j;

    for (i = 0; i <= N; i++)
    {
        for (j = 0; j <= W; j++)
        {
            if(i==0 || j==0)
                k[i][j]=0;
            
            else if(wt[i-1]<=j){
                k[i][j]= max(k[i-1][j],val[i-1]+k[i-1][j-wt[i-1]]);
            }
            else
            {
                k[i][j]=k[i-1][j];
            }
            
        }
        
    }

/****Print DP table
    for(i=0;i<=N;i++){
        for(j=0;j<=W;j++){
            cout<<k[i][j]<<" ";
        }
        cout<<"--"<<endl;
    }
****/
    return k[N][W];
}

File: algorithms/0-1 Knapsack Problem/knapsack.scala
object Knapsack extends App {

  val inputValues = List(60, 100, 120)
  val inputWeights = List(10, 20, 30)
  val inputCapacityWeight = 50
  println(s"Knapsack value is ${knapsack(inputCapacityWeight, inputWeights , inputValues, inputValues.length)}")

  def knapsack(capacityWeight: Int, weights: List[Int], values: List[Int], n: Int): Int = {
    if (n == 0 || capacityWeight == 0) return 0

    val currentWeight = weights(n - 1)

    if (currentWeight > capacityWeight) {
      knapsack(capacityWeight, weights, values, n - 1)
    } else {
      max(values(n - 1) + knapsack(capacityWeight - currentWeight, weights, values, n - 1), knapsack(capacityWeight, weights, values, n - 1))
    }
  }

  private def max(a: Int, b: Int): Int = if (a > b) a else b

}
